lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -148,110 +379,92 @@ LogicalResult triton::gpu::partitionLoop(scf::ForOp loop) {
-      capture = alloc;
-    }
+  SmallVector<int32_t> numWarps(numPartitions, lookupNumWarps(loop));
+  auto wgOp = topBuilder.create<nvws::WarpGroupOp>(resultTypes, numWarps,
+                                                   numPartitions);
 
-    explicitCaptures.push_back(capture);
+  SmallVector<WarpGroupBuilder> builders;
+  for (Region &region : wgOp.getPartitionRegions()) {
+    auto partitionId = builders.size();
+    builders.push_back(
+        WarpGroupBuilder::atBlockEnd(&region.emplaceBlock(), partitionId));
   }
 
-  // Clone the ops into each region in topological order.
-  opsToClone = topologicalSort(opsToClone);
-  for (auto [i, region] : llvm::enumerate(wsOp.getPartitionRegions())) {
-    OpBuilder &b = builders[i];
-    IRMapping &mapping = mappings[i];
-    for (Operation *op : opsToClone) {
-      Value copy = b.clone(*op, mapping)->getResult(0);
-      mapping.map(op->getResult(0), copy);
-      replaceAllUsesInRegionWith(op->getResult(0), copy, *region);
+  cloneForOp(loop, builders, schedule);
+
+  for (auto [b, region, partition] : llvm::zip(
+           builders, wgOp.getPartitionRegions(), schedule.getPartitions())) {
+    auto newForOp = cast<scf::ForOp>(region.front().front());
+    auto outputs = newForOp.getResults();
+    b.setInsertionPointAfter(newForOp);
+
+    if (b.partitionId == 0) {
+      b.create<nvws::WarpGroupYieldOp>(wgOp.getLoc(), outputs);
+    } else {
+      // Tensor results computed by non-default partitions are communicated back
+      // via SMEM.
+      // The calls to getLoopVarIndicesToKeep and isTensorResultComputedBy
+      // below are unnecessary if we can encode the partition index and the
+      // corresponding result tensor index of newForOp in
+      // LoopVarCategory::TensorResultFromOtherPartition. In the absence of such
+      // language support, we end up computing the same information multiple
+      // times.
+      auto [_, reverseIndices] =
+          getLoopVarIndicesToKeep(loop, &partition, schedule);
+      for (size_t i = 0; i < loop.getNumRegionIterArgs(); ++i) {
+        if (loopVarCategories[i] ==
+                LoopVarCategory::TensorResultFromOtherPartition &&
+            isTensorResultComputedBy(loop, i, &partition, schedule)) {
+          auto result = newForOp.getResult(reverseIndices[i]);
+          b.create<LocalStoreOp>(wgOp.getLoc(), result, tensorResultAllocs[i]);
+        }
+      }
+      b.create<nvws::WarpGroupReturnOp>(wgOp.getLoc());
     }
   }
 
-  // Replace the leaves with explicit captures.
-  wsOp->insertOperands(wsOp.getNumOperands(), explicitCaptures);
-  for (Region *region : wsOp.getPartitionRegions()) {
-    for (Value capture : explicitCaptures) {
-      BlockArgument arg =
-          region->addArgument(capture.getType(), capture.getLoc());
-      replaceAllUsesInRegionWith(capture, arg, *region);
+  topBuilder.setInsertionPointAfter(wgOp);
+
+  for (auto [i, res] : llvm::enumerate(loop.getResults())) {
+    if (!res.use_empty()) {
```

Comment:
```suggestion
    if (res.use_empty())
      continue;
```

nit: prefer early exit